homeassistant/components/feedreader/__init__.py

```diff
@@ -46,107 +49,115 @@
     max_entries: int = config[DOMAIN][CONF_MAX_ENTRIES]
-    data_file = hass.config.path(f"{DOMAIN}.pickle")
-    storage = StoredData(data_file)
+    old_data_file = hass.config.path(f"{DOMAIN}.pickle")
+    storage = StoredData(hass, old_data_file)
+    await storage.async_setup()
     feeds = [
-        FeedManager(url, scan_interval, max_entries, hass, storage) for url in urls
+        FeedManager(hass, url, scan_interval, max_entries, storage) for url in urls
     ]
-    return len(feeds) > 0
+
+    for feed in feeds:
+        feed.async_setup()
+
+    return True
 
 
 class FeedManager:
     """Abstraction over Feedparser module."""
 
     def __init__(
         self,
+        hass: HomeAssistant,
         url: str,
         scan_interval: timedelta,
         max_entries: int,
-        hass: HomeAssistant,
         storage: StoredData,
     ) -> None:
         """Initialize the FeedManager object, poll as per scan interval."""
+        self._hass = hass
         self._url = url
         self._scan_interval = scan_interval
         self._max_entries = max_entries
         self._feed: feedparser.FeedParserDict | None = None
-        self._hass = hass
         self._firstrun = True
         self._storage = storage
         self._last_entry_timestamp: struct_time | None = None
-        self._last_update_successful = False
         self._has_published_parsed = False
         self._has_updated_parsed = False
         self._event_type = EVENT_FEEDREADER
         self._feed_id = url
-        hass.bus.listen_once(EVENT_HOMEASSISTANT_START, lambda _: self._update())
-        self._init_regular_updates(hass)
+
+    @callback
+    def async_setup(self) -> None:
+        """Set up the feed manager."""
+        self._hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, self._async_update)
+        async_track_time_interval(
+            self._hass, self._async_update, self._scan_interval, cancel_on_shutdown=True
+        )
 
     def _log_no_entries(self) -> None:
         """Send no entries log at debug level."""
         _LOGGER.debug("No new entries to be published in feed %s", self._url)
 
-    def _init_regular_updates(self, hass: HomeAssistant) -> None:
-        """Schedule regular updates at the top of the clock."""
-        track_time_interval(
-            hass,
-            lambda now: self._update(),
-            self._scan_interval,
-            cancel_on_shutdown=True,
-        )
-
-    @property
-    def last_update_successful(self) -> bool:
-        """Return True if the last feed update was successful."""
-        return self._last_update_successful
+    async def _async_update(self, _: datetime | Event) -> None:
+        """Update the feed and publish new entries to the event bus."""
+        last_entry_timestamp = await self._hass.async_add_executor_job(self._update)
+        if last_entry_timestamp:
+            self._storage.async_put_timestamp(self._feed_id, last_entry_timestamp)
```

Comment:
We only need to update storage if there's a new entry timestamp after the feed update.